prover/src/gpu.rs

```diff
@@ -0,0 +1,284 @@
+//! This module contains GPU acceleration logic for Apple Silicon devices. For now the
+//! logic is limited to GPU accelerating RPO 256 trace commitments.
+use crate::{ExecutionProver, WinterProofOptions};
+use air::{FieldElement, PublicInputs};
+use elsa::FrozenVec;
+use gpu_poly::{
+    plan::{gen_rpo_merkle_tree, GpuRpo256RowMajor},
+    utils::page_aligned_uninit_vector,
+};
+use log::debug;
+use pollster::block_on;
+use processor::crypto::RandomCoin;
+use processor::crypto::{Rpo256, RpoDigest};
+use processor::math::{fft, Felt};
+use processor::ExecutionTrace;
+use std::time::Instant;
+use winter_prover::matrix::{get_evaluation_offsets, Segment};
+use winter_prover::{crypto::MerkleTree, ColMatrix, Prover, RowMatrix, StarkDomain};
+
+const RPO_RATE: usize = Rpo256::RATE_RANGE.end - Rpo256::RATE_RANGE.start;
+
+/// Wraps an [ExecutionProver] and provides GPU acceleration for building Rpo256 trace commitments.
+pub(crate) struct GpuRpoExecutionProver<R>(pub ExecutionProver<Rpo256, R>)
+where
+    R: RandomCoin<BaseField = Felt, Hasher = Rpo256>;
+
+impl<R> Prover for GpuRpoExecutionProver<R>
+where
+    R: RandomCoin<BaseField = Felt, Hasher = Rpo256>,
+{
+    type Air = <ExecutionProver<Rpo256, R> as Prover>::Air;
+    type BaseField = Felt;
+    type Trace = <ExecutionProver<Rpo256, R> as Prover>::Trace;
+    type HashFn = Rpo256;
+    type RandomCoin = R;
+
+    fn options(&self) -> &WinterProofOptions {
+        self.0.options()
+    }
+
+    fn get_pub_inputs(&self, trace: &ExecutionTrace) -> PublicInputs {
+        self.0.get_pub_inputs(trace)
+    }
+
+    /// Computes a low-degree extension (LDE) of the provided execution trace over the specified
+    /// domain and builds a commitment to the extended trace.
+    ///
+    /// The extension is performed by interpolating each column of the execution trace into a
+    /// polynomial of degree = trace_length - 1, and then evaluating the polynomial over the LDE
+    /// domain.
+    ///
+    /// Trace commitment is computed by hashing each row of the extended execution trace, and then
+    /// building a Merkle tree from the resulting hashes.
+    ///
+    /// Interpolations and evaluations are computed on the CPU while hashes are simultaneously
+    /// computed on the GPU:
+    ///
+    /// ```text
+    ///        ──────────────────────────────────────────────────────
+    ///               ┌───┐   ┌────┐   ┌───┐   ┌────┐   ┌───┐
+    ///  CPU:   ... ──┤fft├─┬─┤ifft├───┤fft├─┬─┤ifft├───┤fft├─┬─ ...
+    ///               └───┘ │ └────┘   └───┘ │ └────┘   └───┘ │
+    ///        ╴╴╴╴╴╴╴╴╴╴╴╴╴┼╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴┼╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴┼╴╴╴╴╴╴
+    ///                     │ ┌──────────┐   │ ┌──────────┐   │
+    ///  GPU:               └─┤   hash   │   └─┤   hash   │   └─ ...
+    ///                       └──────────┘     └──────────┘
+    ///        ────┼────────┼────────┼────────┼────────┼────────┼────
+    ///           t=n     t=n+1    t=n+2     t=n+3   t=n+4    t=n+5
```

Comment:
If I understood the code correctly, this picture is not entirely accurate: right now we first interpolate all columns (which I think is fine) and only then start evaluating/hashing them in an interleaved manner.

Reply:
My reply below should answer this but I think this diagram is accurate. The diagram just doesn't represent the quantities of each box.